### ISSUE TITLE
ITP implementation beaten by Roots.jl 

### ISSUE THREAD

**Issue by tecosaur** [SciML/NonlinearSolve.jl#501]
Hello!

I wrote the original code that was developed into the current ITP implementation (#184) and recently for fun thought I'd compare it to what's in Roots.jl using the example problem in https://docs.sciml.ai/SciMLBenchmarksOutput/stable/IntervalNonlinearProblem/simpleintervalrootfind/.

Curiously, it seems like the Roots.jl implementation is ~15% faster. Adding an extra `Roots.ITP()` argument to `find_zero`, I find that (using `Chairmarks` and a Zen2 chip):

```
julia> @b froots(out, levels, (0.0, 2.0))
81.623 ms (100000 allocs: 15.259 MiB, 10.37% gc time)

julia> @b f(out, levels, (0.0, 2.0))
94.051 ms
```

From this, I suspect there's some performance on the table with the current implementation. I don't have time to dig into this myself, but thought I might as well let you folk know :slightly_smiling_face:.

**Comment by ChrisRackauckas** [comment]
Huh interesting... and there's an alloc there? We should definitely dive in.

**Comment by oscardssmith** [comment]
The allocation is in the Roots.jl version. The surprising thing is that despite that, Roots.jl is faster than NonlinearSolve,

**Comment by oscardssmith** [comment]
Interestingly, this doesn't reproduce for me. I'm seeing better results on NonlinearSolve...

**Comment by oscardssmith** [comment]
That said, https://github.com/SciML/NonlinearSolve.jl/pull/547 will save a couple ns

**Comment by tecosaur** [comment]
> The allocation is in the Roots.jl version. The surprising thing is that despite that, Roots.jl is faster than NonlinearSolve,

Indeed, that was exactly what caught my eye.

> Interestingly, this doesn't reproduce for me. I'm seeing better results on NonlinearSolve... @tecosaur what versions are you running?

Let's see, that comment was from late last year so I suspect 1.10 and it would have been on an x86_64 Linux system. I'm guessing from the difference you see now that compiler changes may have helped the ITP implementation here.

**Comment by oscardssmith** [comment]
Yeah, that's plausible. The code is fairly textbook, and is relying on the compiler to do some fixes  (e.g. replace division by 2 with multiplication by .5. In any case, since this aparently no longer reproduces, I think it's safe to close.

**Comment by ChrisRackauckas** [comment]
I also cannot reproduce:

```julia
using Roots, BracketingNonlinearSolve, BenchmarkTools

const N = 100_000;
levels = 1.5 .* rand(N);
out = zeros(N);
myfun(x, lv) = x * sin(x) - lv
function froots(out, levels, u0)
    for i in 1:N
        out[i] = find_zero(x->myfun(x, levels[i]), u0)
    end
end

@btime froots(out, levels, (0, 2)) # 182.550 ms (100000 allocations: 7.63 MiB)

function f(out, levels, u0)
    for i in 1:N
        out[i] = solve(IntervalNonlinearProblem{false}(IntervalNonlinearFunction{false}(myfun),
                u0, levels[i]), ITP()).u
    end
end

@btime f(out, levels, (0.0, 2.0)) # 110.998 ms (0 allocations: 0 bytes)
```

So I'll call this case closed.

**Comment by tecosaur** [comment]
Brilliant, glad to see everything's working as expected now 🙂
